### ISSUE TITLE
RealCall possible response leak?

### ISSUE THREAD

**Issue by Tolriq** [square/okhttp#4583]
I'm trying to understand why I have a what seems impossible very rare  LeakedClosableViolation error when doing lots of call.enqueue cancellations.

I'm sure that I close everything everywhere possible.

While looking at OkHTTP source I saw this: 
https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/RealCall.java#L209

I don't know all the internals but it seems there can be a racy condition here. Where the response is correctly get but call is cancelled just before the isCanceled() test. Leading to the leak as the response is never closed since never passed back to the caller.

cc: @swankjesse 

StrictMode log:
```
2019-02-01 12:38:45.968 4693-4777/org.leetzone.android.yatsewidgetfree.debug D/StrictMode: StrictMode policy violation: android.os.strictmode.LeakedClosableViolation: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
        at android.os.StrictMode$AndroidCloseGuardReporter.report(StrictMode.java:1786)
        at dalvik.system.CloseGuard.warnIfOpen(CloseGuard.java:264)
        at java.lang.reflect.Method.invoke(Native Method)
        at okhttp3.internal.platform.AndroidPlatform$CloseGuard.warnIfOpen(AndroidPlatform.java:355)
        at okhttp3.internal.platform.AndroidPlatform.logCloseableLeak(AndroidPlatform.java:162)
        at okhttp3.ConnectionPool.pruneAndGetAllocationCount(ConnectionPool.java:271)
        at okhttp3.ConnectionPool.cleanup(ConnectionPool.java:211)
        at okhttp3.ConnectionPool$1.run(ConnectionPool.java:60)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.Throwable: Explicit termination method 'response.body().close()' not called
        at dalvik.system.CloseGuard.open(CloseGuard.java:221)
        at java.lang.reflect.Method.invoke(Native Method)
        at okhttp3.internal.platform.AndroidPlatform$CloseGuard.createAndOpen(AndroidPlatform.java:343)
        at okhttp3.internal.platform.AndroidPlatform.getStackTraceForCloseable(AndroidPlatform.java:158)
        at okhttp3.RealCall.captureCallStackTrace(RealCall.java:115)
        at okhttp3.RealCall.enqueue(RealCall.java:124)
        at com.genimee.android.utils.http.OkHttpWorkerPool.await(OkHttpWorkerPool.kt:92)
        at com.genimee.android.utils.http.OkHttpWorkerPool.executeTask(OkHttpWorkerPool.kt:72)
        at com.genimee.android.utils.http.OkHttpWorkerPool$startWorkers$$inlined$repeat$lambda$1.invokeSuspend(OkHttpWorkerPool.kt:45)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
```

**Comment by swankjesse** [comment]
I don't think it's going to be the line you referenced; that only occurs if the executor service gets shut down.

If you have a lead on the cause we might be able to land something before the release. Otherwise there's always another release around the corner.

**Comment by Tolriq** [comment]
@swankjesse The file was modified after the report ;)

The line is now https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/RealCall.java#L211

```
        Response response = getResponseWithInterceptorChain();
        !!! Possible race condition here !!!!
        if (retryAndFollowUpInterceptor.isCanceled()) {
          signalledCallback = true;
          responseCallback.onFailure(RealCall.this, new IOException("Canceled"));
        } else {
          signalledCallback = true;
          responseCallback.onResponse(RealCall.this, response);
        }
}
```

**Comment by swankjesse** [comment]
Worth a look!

**Comment by Tolriq** [comment]
Thanks, and actually my remark about 3.13 was more about the fear that since it's not "security" when 3.13 is out such fix would not make it to 3.12.

For my knowledge, since race conditions is something I totally fail at reproducing / testing is there some specific things to know / tools to help with that?

**Comment by Tolriq** [comment]
@JakeWharton continuing here to try to repro.

I'm facing an issue as can't get StrictMode to trigger in tests :(
How can I test if a response / body is closed ? I can't find a way.

**Comment by swankjesse** [comment]
You might need to induce garbage collection to trigger the bug.

**Comment by swankjesse** [comment]
I recently rearranged a lot of code in RetryAndFollowUpInterceptor, and that _might_ have improved things.

**Comment by Tolriq** [comment]
Not sure I'm a fan of might :(

Took me time to find the 90% sure source, and with that refactor I'm now sure this won't be fixed in 3.12.x
I can't migrate to 3.13 soon so I'm quite stuck.

Would be nice to help build proper tests for the future to have a way to check that a response was closed can't rely on Android Strict mode to detect leaks for a library that is not only Android even if I can get it work in unit tests.

@swankjesse  Anyway new code still have the exact same issue:

https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/RealCall.java#L192

```
Response response = getResponseWithInterceptorChain();
!!!! Race condition here another thread can cancel the call exactly here !!!!!
if (transmitter.isCanceled()) throw new IOException("Canceled");
```

Proper code should be

```
Response response = getResponseWithInterceptorChain();
if (transmitter.isCanceled()) {
   if (response != null && response.body != null) response.close();
   throw new IOException("Canceled");
}
```
